---
title: "2218_Yilong_mRNA_stressGranules_DESeq2"
output: html_notebook
toc: true
---

## Sample sheet

sample name     note    Barcode note-1
1       total RNA of mock-1     22L005998       collected at 20200215
2       total RNA of mock-2     22L005999       collected at 20200215
3       total RNA of mock-3     22L006000       collected at 20200218
4       total RNA of arsenite-1 22L006001       collected at 20200215
5       total RNA of arsenite-2 22L006002       collected at 20200218
6       total RNA of arsenite-3 22L006003       collected at 20200218
7       total RNA of HS-1       22L006004       collected at 20200218
8       total RNA of HS-2       22L006005       collected at 20200218
9       total RNA of HS-3       22L006006       collected at 20200218
10      total RNA of NaCl-1     22L006007       collected at 20200215
11      total RNA of NaCl-2     22L006008       collected at 20200218
12      total RNA of NaCl-3     22L006009       collected at 20200218
13      total RNA of 4sUV-1     22L006010       collected at 20200215
14      total RNA of 4sUV-2     22L006011       collected at 20200215
15      total RNA of 4sUV-3     22L006012       collected at 20200218
16      SG RNA of arsenite-1    22L006013       collected at 20200215
17      SG RNA of arsenite-2    22L006014       collected at 20200218
18      SG RNA of arsenite-3    22L006015       collected at 20200218
19      SG RNA of HS-1  22L006016       collected at 20200218
20      SG RNA of HS-2  22L006017       collected at 20200218
21      SG RNA of HS-3  22L006018       collected at 20200218
22      SG RNA of  NaCl-1       22L006019       collected at 20200215
23      SG RNA of  NaCl-2       22L006020       collected at 20200218
24      SG RNA of  NaCl-3       22L006021       collected at 20200218
25      SG RNA of UV-1  22L006022       collected at 20200215
26      SG RNA of UV-2  22L006023       collected at 20200215
27      SG RNA of UV-3  22L006024       collected at 20200218



```{r, install-libraries}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#	install.packages("BiocManager")

# Install eisaR
BiocManager::install("eisaR")
BiocManager::install("GenomicFeatures")
BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")   # see 'annotation' to figure out which object you need
BiocManager::install("QuasR")

#BiocManager::install(c("tidyverse", "apeglm", "ashr"))
```

```{r, load-libraries}
suppressPackageStartupMessages(library(eisaR))
suppressPackageStartupMessages(library(GenomicFeatures))
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))
suppressPackageStartupMessages(library(QuasR))
```
### 1. Prepare gene annotation

### Load the data
```{r, annotations, message=FALSE, warning=FALSE, include=FALSE}
# view available annotations to know what to install in step 1
#pkgs <- c(BiocManager::available("TxDb"), BiocManager::available("EnsDb"))
#pkgs

# load the object:
txHu <- TxDb.Hsapiens.UCSC.hg38.knownGene



# or create a TxDb or EnsDb out of alternative sources of gene annotations
#for creating TxDb obj: makeTxDb in the GenomicFeatures package
#for creating EnsDb ojb: makeEnsembldbPackage in the ensembldb package

#txHu2 <- makeTxDbFromGFF(file = "references/Mus_musculus.GRCm38.97.gtf", format = "gtf")



# or load the created object, eg a TxDb object as a file:
#txdbFile <- system.file("extdata", "hg19sub.sqlite", package = "eisaR")
#txdb <- AnnotationDbi::loadDb(txdbFile)
```

```{r, explore-TxDb-obj}
keytypes(txHu)
columns(txHu)

# NOT WORKING
# keyList <- ensemblAnnot$GeneID[ensemblAnnot$Symbol=="Wap"]
# select(txMm, 
# 	   keys=keyList,
# 	   keytype = "GENEID",
# 	   columns=c("TXNAME", "TXCHROM", "TXSTART", "TXEND", "TXSTRAND", "TXTYPE")
# 	  )

# extract filtered exonic and gene body regions
regS <- eisaR::getRegionsFromTxDb(txdb = txHu, strandedData = TRUE)
regU <- eisaR::getRegionsFromTxDb(txdb = txHu, strandedData = FALSE)

lengths(regS)
lengths(regU)
# the filtering procedure removes slightly more genes for unstranded data (strandedData = FALSE), as overlapping genes cannot be discriminated even if they reside on opposite strands

regS$genebodies
regS$exons
```



```{r, extract-annot-into-files}
#Export the obtained regions into files. This may be useful if you plan to align and/or quantify reads outside of R. For example, you can use rtracklayer to export the regions in regS into .gtf files:

# library(rtracklayer)
# export(regS$exons, "hg19sub_exons_stranded.gtf")
# export(regS$genebodies, "hg19sub_genebodies_stranded.gtf")
```





### 2. Quantify RNA-seq alignments into exons and introns

#### 2.1. Align reads
Align the reads to genome (hg19sub.fa) using qAlign. 

```{r, align-reads}
# create the sample sheet for paired-end sequencing:
path_inDir <- "/data/processing2/balan_Yilong/project_2297_45nG7vuc_mRNA-seq"
orig_ss <- read.delim(paste0(path_inDir, "/", "sample_sheets/corrected_sampleSheet_switch_UL3_to_UH2.tsv"), header = TRUE, sep='\t', na.strings="NA")
  
fastq_location <- paste0(path_inDir, "/", "fastq")

FileName1 <- list.files(path = fastq_location, pattern = '_R1.fastq.gz', all.files = FALSE,
           full.names = TRUE, recursive = FALSE, 
           ignore.case = FALSE, include.dirs = FALSE, no.. = TRUE)

FileName2 <- list.files(path = fastq_location, pattern = '_R2.fastq.gz', all.files = FALSE,
           full.names = TRUE, recursive = FALSE, 
           ignore.case = FALSE, include.dirs = FALSE, no.. = TRUE)

SampleName <- substring(basename(FileName1),1, nchar(basename(FileName1))-12)
#sample_names <- all_sample_names[seq(1, length(all_sample_names), 2)]

sampleSheet_fastq <- data.frame(FileName1, FileName2, SampleName) 
#################

bam_location <- paste0(path_inDir, "/", "bams")

FileName <- list.files(path = bam_location, pattern = '.bam', all.files = FALSE,
           full.names = TRUE, recursive = FALSE, 
           ignore.case = FALSE, include.dirs = FALSE, no.. = TRUE)

SampleName <- substring(basename(FileName),1, nchar(basename(FileName))-4)
#sample_names <- all_sample_names[seq(1, length(all_sample_names), 2)]

sampleSheet_bam <- data.frame(FileName, SampleName) 
##################

hu_genome = read.delim("/data/repository/organisms/GRCh38_ensembl/genome_fasta/genome.fa")

proj1 <- qAlign(sampleSheet_fastq, hu_genome)

write.table(alignments(proj1)$hu_genome, sampleSheet_bam, sep="\t", row.names=FALSE)
```


### Working only with bam files

Make two sequential calls to qAlign. 

- First, qAlign is called with the orignial sample file (sampleFile1) that lists the raw sequence files

- subsequently with a second sample file (sampleFile2) that lists the bam files generated in the first call.

 
```{r, check_input}


sampleFile1 <- "samples_fastq.txt"
sampleFile2 <- "samples_bam.txt"

proj1 <- qAlign(sampleFile1, genomeFile)

write.table(alignments(proj1)$genome, sampleFile2, sep="\t", row.names=FALSE)

proj2 <- qAlign(sampleFile2, genomeFile)

```
############################

```{r, factoring}
dfmeta_all$condition <- factor(dfmeta_all$condition) 
dfmeta_all$group <- factor(dfmeta_all$group) 
```

```{r, design}
#create design
design <- ~condition
```


```{r, final_check_create_dds_obj}
#check if the col and rows are in the same order:
#all(rownames(dfmeta_all) == colnames(dfcounts_gene))   ## > FALSE
all(rownames(dfmeta_all) == colnames(dfcounts_gene[,rownames(dfmeta_all)]))

#make dds obj - AND rearrange the columns from the counts df to correspond to the ones in the metadata
dds <- DESeq2::DESeqDataSetFromMatrix(countData = dfcounts_gene[,rownames(dfmeta_all)], colData = dfmeta_all, design = design)
################################
```

```{r, plot_PCA_batch}
#png("./manual_PCA.png", width=1600, height=800, units = "px", bg = "white")

rld_PCA <- plotPCA(rlog(dds, blind = TRUE), intgroup="condition", returnData=TRUE)
percentVar <- round(100 * attr(rld_PCA, "percentVar"), 1)   # rounded values for each axis of PC variance
ggplot(rld_PCA, aes(PC1, PC2, color=condition)) +
  geom_point(size=3) +
  scale_shape_manual(values=c(3, 1, 2))+    # change shapes of symbols
  scale_size_manual(values=c(8, 8, 8))+     # change sizes of symbols
  xlab(paste0("PC1: ",percentVar[1])) +
  ylab(paste0("PC2: ",percentVar[2])) 

#dev.off()
```


```{r, DESeq2_batch}
dds<- DESeq2::DESeq(dds)
plotDispEsts(dds)
```

```{r, results}
fdr=0.05
lfc_filter=2
#resultsNames(dds)          # get names of contrasts
#dfmeta_all$condition       # each condition

## We want these comparisons: mock vs (4UH, 4UL) and 4UH vs 4UL
contrast_tot <- list(c("condition", levels(dfmeta_all$condition)[3], levels(dfmeta_all$condition)[4]))

#DE_gene_totRNA_arsenite_vs_totRNA_mock <- DESeq2::results(dds_batch, contrast=contrast_tot[[1]], independentFiltering=TRUE, alpha=0.05, lfcThreshold=lfc_filter, pAdjustMethod="BH", parallel=TRUE)

#DE_gene_totRNA_arsenite_vs_totRNA_mock_ashr <- lfcShrink(dds_batch, contrast=contrast_tot[[1]], res=DE_gene_totRNA_arsenite_vs_totRNA_mock, type="ashr")

#DE_gene_totRNA_arsenite_vs_totRNA_mock_apeglm <- lfcShrink(dds_batch, coef="condition_totRNA_arsenite_vs_totRNA_mock", res=DE_gene_totRNA_arsenite_vs_totRNA_mock, type="apeglm", lfcThreshold=lfc_filter)

for (a in contrast_tot){
  print(paste0("DE_gene_", a[2], "_vs_", a[3]) )

  assign(paste0("DE_gene_", a[2], "_vs_", a[3]), 
         DESeq2::results(dds, contrast=a, alpha=0.05, pAdjustMethod="BH", parallel=TRUE) )
  
  assign(paste0("DE_gene_", a[2], "_vs_", a[3], "_ashr"), 
         DESeq2::lfcShrink(dds, contrast=a, type="ashr", 
         res=get(paste0("DE_gene_", a[2], "_vs_", a[3] ) ) 
         ) 
  )
}


# drawLines <- function() abline(h=c(-2,2),col="dodgerblue",lwd=2)
#plotMA(dds, lfcThreshold=lfc_filter); drawLines()


#################Stopped here######################



output <- list(dds = dds, ddr = DE_gene_treat_4sU_puroHigh_vs_treat_4sU_puroLow, ddr_shrunk = DE_gene_treat_4sU_puroHigh_vs_treat_4sU_puroLow)





DEseqout <- DESeq_basic(countData = '../featureCounts/counts.tsv', colData = coldata, design =d)
```