---
title: "2218_mutation_analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r pressure, echo=FALSE}
path_dir="/data/manke/processing/balan/Yilong_SG_data/2022.03_DE_genes_transcript/SNP_TtoC_mutation_analysis/bcftools_call_mut01"
path_unfiltered_query="/data/manke/processing/balan/Yilong_SG_data/2022.03_DE_genes_transcript/SNP_TtoC_mutation_analysis/bcftools_call_mut01/unfiltered_data.txt"
path_filtered_query="/data/manke/processing/balan/Yilong_SG_data/2022.03_DE_genes_transcript/SNP_TtoC_mutation_analysis/bcftools_call_mut01/filtered_data/filtered_data.txt"

# find out what type of data you have and then load the file faster:
sampleData <- read.delim(path_unfiltered_query, header = TRUE, nrows = 5)
classes <- sapply(sampleData, class)

unfiltered_query <- read.delim(path_unfiltered_query, header = TRUE, sep='\t', na.strings="NA", colClasses=classes)
filtered_query <- read.delim(path_filtered_query, header = TRUE, sep='\t', na.strings="NA", colClasses=classes)

head(unfiltered_query)
head(filtered_query)
```

```{r bcftools_stats}
# get mean for the repeats
#results=data.frame(apply(  array(as.matrix(df1[,-1]), c(nrow(df1), 3, ncol(df1)/3)  ), 3, rowMeans)  )
    # array( df, c(12, 3, 9))                       - create an array of 9 df made each of 3-columns data frames
    # apply( array( df, c(12, 3, 9)), 3, rowMeans)  - calculate the rowMeans for each df made of 3-columns
    # apply( matrix,                  1/2,  f): input is a matrix. output is a vector, where element i is f(row/col i of the matrix)
#results=cbind(df1$Gene, results)

results_unfilter <- data.frame(apply(array(as.matrix(unfiltered_query[,-1]), c(nrow(unfiltered_query),3, ncol(unfiltered_query)/3)),3, rowMeans))
results_unfilter <- cbind(unfiltered_query$mutation, results_unfilter)
colnames(results_unfilter) <- c ('mutation', "totRNA_mock","totRNA_arsenite", "totRNA_HS", "totRNA_NaCl", "totRNA_UV", "sgRNA_arsenite", "sgRNA_HS", "sgRNA_NaCl", "sgRNA_UV")
unfilter <- results_unfilter[7:nrow(results_unfilter),]

results_filter <- data.frame(apply(array(as.matrix(filtered_query[,-1]), c(nrow(filtered_query),3, ncol(filtered_query)/3)),3, rowMeans))
results_filter <- cbind(filtered_query$mutation, results_filter)
colnames(results_filter) <- c ('mutation', "totRNA_mock","totRNA_arsenite", "totRNA_HS", "totRNA_NaCl", "totRNA_UV", "sgRNA_arsenite", "sgRNA_HS", "sgRNA_NaCl", "sgRNA_UV")
filter <- results_filter[7:nrow(results_filter),]
```



```{r mpileup_unfilter_counts}
#png(paste0(path_dir, "/unfiltered_mpileup_T2C.png"), width=1600, height=800, units = "px", bg = "white")
#library(ggplot2)

#ggplot(results_unfilter, aes(x=mutation))+  
#	geom_bar(stat="identity", position=position_dodge()) +  
#	labs(y="Count", title="SNP counts")

# p <- ggplot(data=unfilter, aes(x=mutation, y=totRNA_mock)) +
#     geom_bar(stat="identity", position=position_dodge())
# p

# re-assign mutation column as rownames
df_unfilter <- unfilter[,-1]
rownames(df_unfilter) <- unfilter[,1]

barplot(as.matrix(df_unfilter), col = c("lightblue", "mistyrose", "lightcyan", "lavender", "red", "cornsilk"),
        legend = rownames(df_unfilter), 
        xlab = "sample", ylab = "mutation counts",
        main = "Raw plot for single nucleotide substitutions (1mut/read) #2218",
        beside = TRUE)
#dev.off()
```

```{r mpileup_filter_counts}
#png(paste0(path_dir, "/filtered_mpileup_T2C.png"), width=1600, height=800, units = "px", bg = "white")

df_filter <- filter[,-1]
rownames(df_filter) <- filter[,1]

barplot(as.matrix(df_filter), col = c("lightblue", "mistyrose", "lightcyan", "lavender", "red", "cornsilk"),
        legend = rownames(df_filter), 
        xlab = "sample", ylab = "mutation counts",
        main = "Single nucleotide substitutions (Q>20, DP >3, 1mut/read) #2218",
        beside = TRUE)
#dev.off()
```
```{r mpileup_unfilter_divSum}
#png(paste0(path_dir, "/unfiltered_mpileup_T2C_divSum.png"), width=1600, height=800, units = "px", bg = "white")
library(dplyr)


df_unfilter_divSum <- df_unfilter %>%
    mutate_if(is.numeric, funs(./sum(.))) 

barplot(as.matrix(df_unfilter_divSum), col = c("lightblue", "mistyrose", "lightcyan", "lavender", "red", "cornsilk"),
        legend = rownames(df_unfilter), 
        xlab = "sample", ylab = "mutation frequency (counts mut/tot mut)",
        main = "Raw plot for single nucleotide substitutions (1mut/read) #2218",
        beside = TRUE)
#dev.off()
```

```{r mpileup_filter_divSum}
png(paste0(path_dir, "/filtered_mpileup_T2C_divSum.png"), width=1600, height=800, units = "px", bg = "white")

df_filter_divSum <- df_filter %>%
    mutate_if(is.numeric, funs(./sum(.))) 

barplot(as.matrix(df_filter_divSum), col = c("lightblue", "mistyrose", "lightcyan", "lavender", "red", "cornsilk"),
        legend = rownames(df_filter), 
        xlab = "sample", ylab = "mutation frequency (counts mut/tot mut)",
        main = "mutation frequency (counts mut/tot mut) (Q>20, DP >3, 1mut/read) #2218",
        beside = TRUE)
dev.off()
```


```{r mpileup_unfilter_covNucl}
#png(paste0(path_dir, "/unfiltered_mpileup_T2C_covNucl.png"), width=1600, height=800, units = "px", bg = "white")

# find out what type of data you have and then load the file faster:
sampleData3 <- read.delim(paste0(path_dir, "/s9_all_grep"), header = FALSE, nrows = 5)
classes3 <- sapply(sampleData3, class)

s9_all_mut <- read.delim(paste0(path_dir, "/s9_all_grep"), header = FALSE, sep='\t', na.strings="NA", colClasses = classes3)


aggregate(.~id1+id2, all_mut, mean)

barplot(as.matrix(df_unfilter_covNucl), col = c("lightblue", "mistyrose", "lightcyan", "lavender", "red", "cornsilk"),
        legend = rownames(df_unfilter), 
        xlab = "sample", ylab = "mutation frequency (counts mut/orig Nucl)",
        main = "Raw plot for single nucleotide substitutions (counts mut/orig Nucl, 1mut/read) #2218",
        beside = TRUE)
#dev.off()
```

```{r mpileup_filter_covNucl}
#png(paste0(path_dir, "/filtered_mpileup_T2C_covNucl.png"), width=1600, height=800, units = "px", bg = "white")



barplot(as.matrix(df_filter_covNucl), col = c("lightblue", "mistyrose", "lightcyan", "lavender", "red", "cornsilk"),
        legend = rownames(df_filter), 
        xlab = "sample", ylab = "mutation frequency (counts mut/orig Nucl)",
        main = "mutation frequency (counts mut/orig Nucl) (Q>20, DP >3, 1mut/read) #2218",
        beside = TRUE)
#dev.off()
```





```{r py_test}
# sub plot_substitutions
# {
#     my ($opts,$id) = @_;
# 
#     my @vals = get_values($opts,$id,'ST');
#     if ( !@vals ) { return; }
# 
#     my $fh  = $$opts{plt_fh};
#     my $img = "substitutions.$id";
# 
#     tprint $fh, "
#             dat = [
#         ";
#     for (my $i=0; $i<@vals; $i++) { my $val=$vals[$i]; tprint $fh, "\t[$i,'$$val[0]',$$val[1]],\n"; }
#     tprint $fh, "]\n";
#     tprint $fh, "
#             if plot_substitutions:
#                 fig = plt.figure(figsize=($$opts{img_width},$$opts{img_height}))
#                 cm  = mpl.cm.get_cmap('autumn')
#                 n = 12
#                 col = []
#                 for i in list(range(n)): col.append(cm(1.*i/n))
#                 ax1 = fig.add_subplot(111)
#                 ax1.bar([row[0] for row in dat], [row[2] for row in dat], color=col)
#                 ax1.set_ylabel('Count')
#                 ax1.ticklabel_format(style='sci', scilimits=(0,0), axis='y')
#                 ax1.set_xlim(-0.5,n+0.5)
#                 plt.xticks([row[0] for row in dat],[row[1] for row in dat],rotation=45)
#                 plt.title('$$opts{title}{$id}')
#                 plt.savefig('$img.png')
#                 if img_fmt != 'png': plt.savefig('$img.' + img_fmt)
#                 plt.close()
#         ";
# }

```