---
title: "2218_Yilong_mRNA_stressGranules_DESeq2"
output: html_notebook
toc: true
---

## Sample sheet

sample name     note    Barcode note-1
1       total RNA of mock-1     22L005998       collected at 20200215
2       total RNA of mock-2     22L005999       collected at 20200215
3       total RNA of mock-3     22L006000       collected at 20200218
4       total RNA of arsenite-1 22L006001       collected at 20200215
5       total RNA of arsenite-2 22L006002       collected at 20200218
6       total RNA of arsenite-3 22L006003       collected at 20200218
7       total RNA of HS-1       22L006004       collected at 20200218
8       total RNA of HS-2       22L006005       collected at 20200218
9       total RNA of HS-3       22L006006       collected at 20200218
10      total RNA of NaCl-1     22L006007       collected at 20200215
11      total RNA of NaCl-2     22L006008       collected at 20200218
12      total RNA of NaCl-3     22L006009       collected at 20200218
13      total RNA of 4sUV-1     22L006010       collected at 20200215
14      total RNA of 4sUV-2     22L006011       collected at 20200215
15      total RNA of 4sUV-3     22L006012       collected at 20200218
16      SG RNA of arsenite-1    22L006013       collected at 20200215
17      SG RNA of arsenite-2    22L006014       collected at 20200218
18      SG RNA of arsenite-3    22L006015       collected at 20200218
19      SG RNA of HS-1  22L006016       collected at 20200218
20      SG RNA of HS-2  22L006017       collected at 20200218
21      SG RNA of HS-3  22L006018       collected at 20200218
22      SG RNA of  NaCl-1       22L006019       collected at 20200215
23      SG RNA of  NaCl-2       22L006020       collected at 20200218
24      SG RNA of  NaCl-3       22L006021       collected at 20200218
25      SG RNA of UV-1  22L006022       collected at 20200215
26      SG RNA of UV-2  22L006023       collected at 20200215
27      SG RNA of UV-3  22L006024       collected at 20200218



```{r, install-libraries}
##if (!requireNamespace("BiocManager", quietly = TRUE))
##	install.packages("BiocManager")

## Install eisaR
# BiocManager::install("eisaR")
# BiocManager::install("GenomicFeatures")
# BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")   # see 'annotation' to figure out which object you need
# BiocManager::install("QuasR")
# BiocManager::install("gtools")
# BiocManager::install("ensembldb")
# delete?
#BiocManager::install("BiocParallel")

##BiocManager::install(c("tidyverse", "apeglm", "ashr"))
```

```{r, load-libraries}
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(eisaR))
suppressPackageStartupMessages(library(GenomicFeatures))
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))  # geneID to symbol conversion and for annotation
suppressPackageStartupMessages(library(QuasR))
suppressPackageStartupMessages(library(gtools))            # to do alphanumeric sort on column names (+ stringAsFactors=F)
suppressPackageStartupMessages(library(ensembldb))
suppressPackageStartupMessages(library(edgeR))
suppressPackageStartupMessages(library(BiocParallel))
```
### 1. Prepare gene annotation

Useless

### Load the data
```{r, annotations, message=FALSE, warning=FALSE, include=FALSE}
# # view available annotations to know what to install in step 1
# #pkgs <- c(BiocManager::available("TxDb"), BiocManager::available("EnsDb"))
# #pkgs
# 
# # load the object:
# txHu <- TxDb.Hsapiens.UCSC.hg38.knownGene
# 
# 
# 
# # or create a TxDb or EnsDb out of alternative sources of gene annotations
# #for creating TxDb obj: makeTxDb in the GenomicFeatures package
# #for creating EnsDb ojb: makeEnsembldbPackage in the ensembldb package
# 
# #txHu2 <- makeTxDbFromGFF(file = "references/Mus_musculus.GRCm38.97.gtf", format = "gtf")
# 
# 
# 
# # or load the created object, eg a TxDb object as a file:
# #txdbFile <- system.file("extdata", "hg19sub.sqlite", package = "eisaR")
# #txdb <- AnnotationDbi::loadDb(txdbFile)
```

#Useless
```{r, explore-TxDb-obj}
# keytypes(txHu)
# columns(txHu)
# 
# # NOT WORKING
# # keyList <- ensemblAnnot$GeneID[ensemblAnnot$Symbol=="Wap"]
# # select(txMm, 
# # 	   keys=keyList,
# # 	   keytype = "GENEID",
# # 	   columns=c("TXNAME", "TXCHROM", "TXSTART", "TXEND", "TXSTRAND", "TXTYPE")
# # 	  )
# 
# # extract filtered exonic and gene body regions
# regS <- eisaR::getRegionsFromTxDb(txdb = txHu, strandedData = TRUE)
# regU <- eisaR::getRegionsFromTxDb(txdb = txHu, strandedData = FALSE)
# 
# lengths(regS)
# lengths(regU)
# # the filtering procedure removes slightly more genes for unstranded data (strandedData = FALSE), as overlapping genes cannot be discriminated even if they reside on opposite strands
# 
# regS$genebodies
# regS$exons
```



```{r, extract-annot-into-files}
#Export the obtained regions into files. This may be useful if you plan to align and/or quantify reads outside of R. For example, you can use rtracklayer to export the regions in regS into .gtf files:

# library(rtracklayer)
# export(regS$exons, "hg19sub_exons_stranded.gtf")
# export(regS$genebodies, "hg19sub_genebodies_stranded.gtf")
```





### NO
#### 2. Quantify RNA-seq alignments into exons and introns

#### 2.1. Align reads
Align the reads to genome (hg19sub.fa) using qAlign. 

```{r, align-reads}
# # create the sample sheet for paired-end sequencing:
# path_inDir <- "/data/processing2/balan_Yilong/project_2297_45nG7vuc_mRNA-seq"
# orig_ss <- read.delim(paste0(path_inDir, "/", "sample_sheets/corrected_sampleSheet_switch_UL3_to_UH2.tsv"), header = TRUE, sep='\t', na.strings="NA")
#   
# fastq_location <- paste0(path_inDir, "/", "fastq")
# 
# FileName1 <- list.files(path = fastq_location, pattern = '_R1.fastq.gz', all.files = FALSE,
#            full.names = TRUE, recursive = FALSE, 
#            ignore.case = FALSE, include.dirs = FALSE, no.. = TRUE)
# 
# FileName2 <- list.files(path = fastq_location, pattern = '_R2.fastq.gz', all.files = FALSE,
#            full.names = TRUE, recursive = FALSE, 
#            ignore.case = FALSE, include.dirs = FALSE, no.. = TRUE)
# 
# SampleName <- substring(basename(FileName1),1, nchar(basename(FileName1))-12)
# #sample_names <- all_sample_names[seq(1, length(all_sample_names), 2)]
# 
# sampleSheet_fastq <- data.frame(FileName1, FileName2, SampleName) 
# #################
# 
# bam_location <- paste0(path_inDir, "/", "bams")
# 
# FileName <- list.files(path = bam_location, pattern = '.bam', all.files = FALSE,
#            full.names = TRUE, recursive = FALSE, 
#            ignore.case = FALSE, include.dirs = FALSE, no.. = TRUE)
# 
# SampleName <- substring(basename(FileName),1, nchar(basename(FileName))-4)
# #sample_names <- all_sample_names[seq(1, length(all_sample_names), 2)]
# 
# sampleSheet_bam <- data.frame(FileName, SampleName) 
# ##################
# 
# hu_genome = read.delim("/data/repository/organisms/GRCh38_ensembl/genome_fasta/genome.fa")
# 
# proj1 <- qAlign(sampleSheet_fastq, hu_genome)
# 
# write.table(alignments(proj1)$hu_genome, sampleSheet_bam, sep="\t", row.names=FALSE)
```

#NO
### Working only with bam files

Make two sequential calls to qAlign. 

- First, qAlign is called with the orignial sample file (sampleFile1) that lists the raw sequence files

- subsequently with a second sample file (sampleFile2) that lists the bam files generated in the first call.

 
```{r, check_input}


# sampleFile1 <- "samples_fastq.txt"
# sampleFile2 <- "samples_bam.txt"
# 
# proj1 <- qAlign(sampleFile1, genomeFile)
# 
# write.table(alignments(proj1)$genome, sampleFile2, sep="\t", row.names=FALSE)
# 
# proj2 <- qAlign(sampleFile2, genomeFile)

```
############################

```{r, factoring}
#dfmeta_all$condition <- factor(dfmeta_all$condition) 
#dfmeta_all$group <- factor(dfmeta_all$group) 
```


## YES

Load exon data and genebody (intron+exon) data.

```{r, load-count-data}
# # create the sample sheet for paired-end sequencing:
path_inDir <- "/data/manke/processing/balan/Yilong_SG_data/2022.03_DE_genes_transcript"
counts_geneBody <- read.delim(paste0(path_inDir, "/analysis_bam_default/", "featureCounts_geneBody/counts.tsv"), header = TRUE, sep='\t', na.strings="NA", row.names = 1, stringsAsFactors = F)
counts_exon <- read.delim(paste0(path_inDir, "/analysis_bam_default/", "featureCounts_defaultExon/counts.tsv"), header = TRUE, sep='\t', na.strings="NA", row.names = 1, stringsAsFactors = F)

## alphanumeric sort of columns
counts_geneBody <- counts_geneBody[,gtools::mixedorder(colnames(counts_geneBody))]
counts_exon <- counts_exon[,gtools::mixedorder(colnames(counts_exon))]

##remove the version nr after each accession number (e.g. "ENSG00000278267.1" to "ENSG00000278267")
  #why doesn't this work?
    #row.names(counts_geneBody) <- gsub("\\..*","",row.names(counts_geneBody))  
    # this would work but I don't understand what is happening:
      # row.names(counts_geneBody) <- make.names(gsub("\\..*","",row.names(counts_geneBody)), unique=TRUE )
  
  # apparently this creates non-unique row names; how many?
    #table(duplicated(gsub("\\..*","",row.names(counts_geneBody))))   
  # I get 45 gene IDs that will become non-unique if I do it this way; which ones? 
    #row.names(counts_geneBody)[duplicated(gsub("\\..*","",row.names(counts_geneBody)))]   
  # all have '_PAR_Y' suffix. Check one of the names:
    #row.names(counts_geneBody[grepl(pattern="ENSG00000169100", row.names(counts_geneBody) ),])
      #  "ENSG00000169100.13"       "ENSG00000169100.13_PAR_Y"   
  # are sequences from Y chromosome and they seem empty
      # all(counts_geneBody[grepl(pattern="_PAR_Y", row.names(counts_geneBody) ),] == 0)    # TRUE
  # are all empty in the 'counts_exon' too?
      # all(counts_exon[grepl(pattern="_PAR_Y", row.names(counts_exon) ),] == 0)            # TRUE

## NOTES:
  # Since by normal definitions there is no Y chromosome in female tissue, PAR_Y quantifications should be ignored (excluding special cases such as intersex syndrome).
  # In male, however, I would suggest making the sum of counts for the X and PAR_Y transcripts. Since alternative splicing may generate different variants of the same "Ensembl-ID" transcript, quantification should account for the total of the isoforms.  
  
  # remove the all the '_PAR_Y' rows because are empty
less_counts_geneBody <- counts_geneBody[!row.names(counts_geneBody) %in% row.names(counts_geneBody[grepl(pattern="_PAR_Y", row.names(counts_geneBody) ),]),]
less_counts_exon <- counts_exon[!row.names(counts_exon) %in% row.names(counts_exon[grepl(pattern="_PAR_Y", row.names(counts_exon) ),]),]

  # remove the version number
  row.names(less_counts_geneBody) <- gsub("\\..*","",row.names(less_counts_geneBody))
  row.names(less_counts_exon) <- gsub("\\..*","", row.names(less_counts_exon))
```

Get intron counts:

```{r, intron-counts}
counts_introns = less_counts_geneBody - less_counts_exon
outdir <- '/intron_analysis_bam/'
dir.create(paste0(path_inDir, outdir), showWarnings = TRUE, recursive = FALSE, mode = "0755")

write.table(counts_introns, file=paste0(path_inDir, outdir, 'counts_introns_from_featureCounts.tsv'), sep="\t", row.names=TRUE)

head(counts_introns)

# create condition factors
treat <- factor(c(rep("totRNA_mock", 3), rep("totRNA_arsenite", 3), rep("totRNA_HS", 3), rep("totRNA_NaCl", 3), rep("totRNA_UV", 3), 
                       rep("sgRNA_arsenite", 3), rep("sgRNA_HS", 3), rep("sgRNA_NaCl", 3), rep("sgRNA_UV", 3)) )
```


```{r, sanity-check-values, eval=FALSE, include=FALSE}
# #sanity checks only for s1 sample:
# #counts_exon[counts_geneBody$s1 == 0,]["ENSG00000278267.1",]
# #dplyr::filter(counts_geneBody,counts_geneBody$s1==0 & counts_exon$s1 != 0)["ENSG00000278267.1",]
# 
# s1_counts_to_zero_in_geneBody <- dplyr::filter(counts_exon,counts_geneBody$s1 == 0 & counts_exon$s1 != 0) # 3 387
# #s1_genes_to_zero_in_geneBody <- row.names(s1_counts_to_zero_in_geneBody)
# 
# s1_counts_less_in_geneBody_than_exon <- dplyr::filter(counts_exon,counts_geneBody$s1 < counts_exon$s1)        # 3 807
# #s1_genes_less_in_geneBody_than_exon <- row.names(s1_counts_less_in_geneBody_than_exon)
# 
# #sort(s1_counts_to_zero_in_geneBody$s1, decreasing = TRUE)
#   #  [1] 78384  4028  3280  2150  2149  2136  2013  1962  1849  1811  1690  1627  1624 
# s1_counts_to_zero_in_geneBody_sort <- s1_counts_to_zero_in_geneBody[order(s1_counts_to_zero_in_geneBody$s1, decreasing=TRUE),]
# s1_counts_less_in_geneBody_than_exon_sort <- s1_counts_less_in_geneBody_than_exon[order(s1_counts_less_in_geneBody_than_exon$s1, decreasing=TRUE),]
# 
# #remove the version nr after each accession number (e.g. "ENSG00000278267.1" to "ENSG00000278267")
# row.names(s1_counts_to_zero_in_geneBody_sort) <- gsub("\\..*","",row.names(s1_counts_to_zero_in_geneBody_sort))
# row.names(s1_counts_less_in_geneBody_than_exon_sort) <- gsub("\\..*","", row.names(s1_counts_less_in_geneBody_than_exon_sort))
# 
# # Convert from ensembl.gene to gene.symbol   
# #keytypes(TxDb.Hsapiens.UCSC.hg38.knownGene)      -doesn't work
# #ENTREZID_TxDb <- keys(TxDb.Hsapiens.UCSC.hg38.knownGene, keytype = "GENEID")
# 
# # works
# #BiocManager::install("EnsDb.Hsapiens.v86")
# #suppressPackageStartupMessages(library(EnsDb.Hsapiens.v86))
# #supportedFilters(EnsDb.Hsapiens.v86)
# 
# s1_genes_to_zero_in_geneBody_sort <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(s1_counts_to_zero_in_geneBody_sort), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
# s1_genes_less_in_geneBody_than_exon_sort <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(s1_counts_less_in_geneBody_than_exon_sort), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
# 
# # the opposite: Convert from gene.symbol to ensembl.gene
# #geneID_s1_to_zero <- ensembldb::select(EnsDb.Hsapiens.v86, keys=geneSymbol_s1_to_zero$SYMBOL, keytype = "SYMBOL", columns = c("SYMBOL","GENEID"))
# 
# write.table(s1_genes_to_zero_in_geneBody_sort$SYMBOL, file=paste0(path_inDir, '/intron_analysis_bam/', 'geneSymbol_s1_to_zero'), sep="\t", row.names=FALSE, col.names=FALSE, quote=FALSE)
# write.table(s1_genes_less_in_geneBody_than_exon_sort$SYMBOL, file=paste0(path_inDir, '/intron_analysis_bam/', 'geneSymbol_s1_less'), sep="\t", row.names=FALSE, col.names=FALSE, quote=FALSE)
#####################################
```

```{r, get-intron-counts-and-set-neg-to-zero}
# # function for looking at genes that have a exon count > geneBody count: (NOT FINISHED-not needed)
# 
# for (a in 1:length(colnames(counts_geneBody))){
#   assign(
#     paste0(colnames(counts_geneBody[1]),"_counts_to_less_in_geneBody_than_exon"), 
#     dplyr::filter(counts_exon, counts_geneBody[1] < counts_exon[1]) )
#   assign(
#     paste0(colnames(counts_geneBody[1]),"_counts_to_less_in_geneBody_than_exon_sort"), 
#     get(paste0(colnames(counts_geneBody[1]),"_counts_to_less_in_geneBody_than_exon"))[order(get(paste0(colnames(counts_geneBody[1]),"_counts_to_less_in_geneBody_than_exon")), decreasing=TRUE),]  )
#   }
# 
# row.names(paste0(colnames(counts_geneBody),"_counts_to_less_in_geneBody_than_exon_sort")) <- gsub("\\..*","", row.names(paste0(colnames(counts_geneBody),"_counts_to_less_in_geneBody_than_exon_sort")))  
# 
# write.table(paste0(colnames(counts_geneBody),"_counts_to_less_in_geneBody_than_exon_sort"), file=paste0(path_inDir, '/intron_analysis_bam/', 'intron_count_from_featureCounts'), sep="\t", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

Eisa doesn't work with negative values. So in this 1st attempt, set all negative values to zero.

```{r, fix-the-negative-values}
counts_introns[counts_introns<0] <- 0
write.table(counts_introns, file=paste0(path_inDir, '/intron_analysis_bam/', 'counts_introns_from_featureCounts_noNeg.tsv'), sep="\t", row.names=TRUE)
```



# run EISA
```{r, eisa-totalRNA}
# # it works only on 2 levels at a time!
# with modelSamples
res_totRNA_ars_vs_mock <- runEISA(  less_counts_exon[c(4:6, 1:3)], counts_introns[c(4:6, 1:3)], droplevels(treat[c(4:6, 1:3)])  )   # drops unused factor levels because EISA only takes 2 at a time

res_totRNA_HS_vs_mock <- runEISA(   less_counts_exon[c(7:9, 1:3)], counts_introns[c(7:9, 1:3)], droplevels(treat[c(7:9, 1:3)])  )   
res_totRNA_NaCl_vs_mock <- runEISA(   less_counts_exon[c(10:12, 1:3)], counts_introns[c(10:12, 1:3)], droplevels(treat[c(10:12, 1:3)])  )   
res_totRNA_UV_vs_mock <- runEISA(   less_counts_exon[c(13:15, 1:3)], counts_introns[c(13:15, 1:3)], droplevels(treat[c(13:15, 1:3)])  )   
```
```{r, eisa-sgRNA}
# with modelSamples
res_sgRNA_HS_vs_ars <- runEISA(   less_counts_exon[c(19:21, 16:18)], counts_introns[c(19:21, 16:18)], droplevels(treat[c(19:21, 16:18)])  )
res_sgRNA_NaCl_vs_ars <- runEISA(   less_counts_exon[c(22:24, 16:18)], counts_introns[c(22:24, 16:18)], droplevels(treat[c(22:24, 16:18)])  )   
res_sgRNA_UV_vs_ars <- runEISA(   less_counts_exon[c(25:27, 16:18)], counts_introns[c(25:27, 16:18)], droplevels(treat[c(25:27, 16:18)])  )   

res_sgRNA_NaCl_vs_HS <- runEISA(   less_counts_exon[c(22:24, 19:21)], counts_introns[c(22:24, 19:21)], droplevels(treat[c(22:24, 19:21)])  )   
res_sgRNA_UV_vs_HS <- runEISA(   less_counts_exon[c(25:27, 19:21)], counts_introns[c(25:27, 19:21)], droplevels(treat[c(25:27, 19:21)])  )   

res_sgRNA_UV_vs_NaCl <- runEISA(   less_counts_exon[c(25:27, 22:24)], counts_introns[c(25:27, 22:24)], droplevels(treat[c(25:27, 22:24)])  )   
```


# run plotEISA
```{r, plotEISA-totalRNA}
# png(paste0(path_inDir, '/intron_analysis_bam/totRNA_comparisons/','eisa_plot_totRNA_LFC.png'), width=1600, height=800, units = "px", bg = "white")
# 
# par(oma = c(3,4,4,1),    # margins for whole plot
#     mar = c(3,4,3.2,1)  # margins for individual plots
#     ) 
# 
# xl <- c(-10,10)
# yl <- c(-8,8)
#  
# layout(matrix(c(1, 2, 3, 4), 2, 2, byrow=TRUE))
# # layout.show(6)
# 
# list_objects=list("res_totRNA_ars_vs_mock", "res_totRNA_HS_vs_mock", "res_totRNA_NaCl_vs_mock", "res_totRNA_UV_vs_mock")
#   
#   
# for (i in 1:length(list_objects) ) {
#   eisaR::plotEISA(get(list_objects[[i]]), maxFDR = 0.05, #minLfc = 2,
#                   xlim=xl, ylim=yl,  
#                   main=list_objects[[i]] , xlab='', ylab='', 
#                   cex.axis=1.5, cex.main=1.5)
# }
# 
# title(main = "EISA - totRNA_intron-exon analysis, FDR < 0.05, LFC > 2",
#       xlab = expression(paste(Delta,'intron (cond1-cond2)')),
#       ylab = expression(paste(Delta,'exon (cond1-cond2)')),
#       outer = TRUE, line = 1,
#       cex.lab=2, cex.main=2)
# 
# dev.off()
```

```{r, ploEISA-sgRNA}
# png(paste0(path_inDir, '/intron_analysis_bam/sgRNA_comparisons/','eisa_plot_sgRNA_LFC_revUV.png'), width=1600, height=800, units = "px", bg = "white")
# 
# par(oma = c(3,4,4,1),    # margins for whole plot
#     mar = c(3,4,3.2,1)  # margins for individual plots
#     )
# 
# xl <- c(-12,12)
# yl <- c(-8,8)
# 
# layout(matrix(c(1, 2, 3, 4, 5, 6), 2, 3, byrow=TRUE))
# # layout.show(6)
# 
# list_objects=list("res_sgRNA_HS_vs_ars", "res_sgRNA_NaCl_vs_ars", "res_sgRNA_UV_vs_ars",
#                   "res_sgRNA_NaCl_vs_HS", "res_sgRNA_UV_vs_HS",
#                   "res_sgRNA_UV_vs_NaCl")
# 
# 
# for (i in 1:length(list_objects) ) {
#   eisaR::plotEISA(get(list_objects[[i]]), maxFDR = 0.05, minLfc = 2,
#                   xlim=xl, ylim=yl,
#                   main=list_objects[[i]] , xlab='', ylab='',
#                   cex.axis=1.5, cex.main=1.5)
#   abline(h = 0, v = 0, col = "gray")
# }
# 
# title(main = "EISA - sgRNA_intron-exon analysis, FDR < 0.05, LFC > 2",
#       xlab = expression(paste(Delta,'intron (cond1-cond2)')),
#       ylab = expression(paste(Delta,'exon (cond1-cond2)')),
#       outer = TRUE, line = 1,
#       cex.lab=2, cex.main=2)
# 
# 
# 
# dev.off()
```

```{r, michael-plots}
# library(rtracklayer)
# gtf_location <- "/data/repository/organisms/GRCh38_ensembl/gencode/release_27/genes.gtf"
# hg38.genes <- rtracklayer::import.gff(gtf_location, feature.type='gene')
# gid2symbol = mcols(hg38.genes)[c('gene_id','gene_name')] %>% deframe
# 
# 
# 
# tab0 <- res_totRNA_ars_vs_mock$tab.ExIn %>% rownames_to_column('gene_id') %>%
#     left_join(as.data.frame( res_totRNA_ars_vs_mock$contrasts) %>% rownames_to_column('gene_id'))
# 
# tab0 <- tab0 %>% mutate(gene_name = gid2symbol[gene_id])
# 
# gene_name[] == gid2symbol['ENSG00000278073.1']
# #names(gid2symbol)[gid2symbol=='SAMD11']
# 
# 
# names(v)[v == "NY"] 
# 
# 
# 
# ggplot(dt0, aes(Introns/Genebody, Exons/Genebody)) + geom_point(aes(color = group, pch = ip), size = 3) +
#   coord_cartesian(xlim = c(0,.5), ylim = c(0,.5)) +
#   geom_abline(slope = 1, intercept = 0) +
#   theme_bw()

















```



```{r, look-at-eisa-totRNA-results}
# res_totRNA_ars_vs_mock
# res_totRNA_HS_vs_mock
# res_totRNA_NaCl_vs_mock
# res_totRNA_UV_vs_mock

# number of genes with significant post-transcriptional regulation
sum(res_totRNA_ars_vs_mock$tab.ExIn$FDR < 0.05)
sum(res_totRNA_HS_vs_mock$tab.ExIn$FDR < 0.05)
sum(res_totRNA_NaCl_vs_mock$tab.ExIn$FDR < 0.05)
sum(res_totRNA_UV_vs_mock$tab.ExIn$FDR < 0.05)


# comparison of contrasts
# diag(cor(res_totRNA_ars_vs_mock$contrasts[ids, ], res_totRNA_ars_vs_mock_o$contrasts[ids, ]))
# plot(res_totRNA_ars_vs_mock$contrasts[ids, 3], res_totRNA_ars_vs_mock_o$contrasts[ids, 3], pch = "*",
#      xlab = "Interaction effects for modelSamples=FALSE",
#      ylab = "Interaction effects for modelSamples=TRUE")

# # comparison of interaction significance
# plot(-log10(res_totRNA_ars_vs_mock$tab.ExIn[ids, "FDR"]), -log10(res_totRNA_ars_vs_mock_o$tab.ExIn[ids, "FDR"]), pch = "*",
#      xlab = "-log10(FDR) for modelSamples=FALSE",
#      ylab = "-log10(FDR) for modelSamples=TRUE")
# abline(a = 0, b = 1, col = "gray")
# legend("bottomright", "y = x", bty = "n", lty = 1, col = "gray")
```


```{r, get-totRNA-results}
#why?
#res_totRNA_ars_vs_mock$designMatrix
# checked the last 2 columns in the vignette (c1.ex and c2.ex) and they are the same

# get results out
out_totRNA_ars_vs_mock <- res_totRNA_ars_vs_mock$tab.ExIn[order(res_totRNA_ars_vs_mock$tab.ExIn$FDR, decreasing=FALSE),]
out_totRNA_HS_vs_mock <- res_totRNA_HS_vs_mock$tab.ExIn[order(res_totRNA_HS_vs_mock$tab.ExIn$FDR, decreasing=FALSE),]
out_totRNA_NaCl_vs_mock <- res_totRNA_NaCl_vs_mock$tab.ExIn[order(res_totRNA_NaCl_vs_mock$tab.ExIn$FDR, decreasing=FALSE),]
out_totRNA_UV_vs_mock<- res_totRNA_UV_vs_mock$tab.ExIn[order(res_totRNA_UV_vs_mock$tab.ExIn$FDR, decreasing=FALSE),]

# change to write to 'totRNA_comparisons' folder
write.table(format(out_totRNA_ars_vs_mock, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_totRNA_ars_vs_mock.tsv'), sep="\t", row.names=TRUE, quote = FALSE)
write.table(format(out_totRNA_HS_vs_mock, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_totRNA_HS_vs_mock.tsv'), sep="\t", row.names=TRUE, quote = FALSE)
write.table(format(out_totRNA_NaCl_vs_mock, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_totRNA_NaCl_vs_mock.tsv'), sep="\t", row.names=TRUE, quote = FALSE)
write.table(format(out_totRNA_UV_vs_mock, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_totRNA_UV_vs_mock.tsv'), sep="\t", row.names=TRUE, quote = FALSE)

#BiocManager::install("EnsDb.Hsapiens.v86")
suppressPackageStartupMessages(library(EnsDb.Hsapiens.v86))
#supportedFilters(EnsDb.Hsapiens.v86)

# Convert from ensembl.gene to gene.symbol   
out_totRNA_ars_vs_mock_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_totRNA_ars_vs_mock), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
out_totRNA_HS_vs_mock_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_totRNA_HS_vs_mock), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
out_totRNA_NaCl_vs_mock_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_totRNA_NaCl_vs_mock), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
out_totRNA_UV_vs_mock_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_totRNA_UV_vs_mock), keytype = "GENEID", columns = c("SYMBOL","GENEID"))

# change to write to 'totRNA_comparisons' folder
write.table(out_totRNA_ars_vs_mock_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_totRNA_ars_vs_mock_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
write.table(out_totRNA_HS_vs_mock_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_totRNA_HS_vs_mock_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
write.table(out_totRNA_NaCl_vs_mock_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_totRNA_NaCl_vs_mock_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
write.table(out_totRNA_UV_vs_mock_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_totRNA_UV_vs_mock_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
```


```{r, get-sgRNA-results}
# get results out

out_sgRNA_HS_vs_ars <- res_sgRNA_HS_vs_ars$tab.ExIn[order(res_sgRNA_HS_vs_ars$tab.ExIn$FDR, decreasing=FALSE),]
out_sgRNA_NaCl_vs_ars <- res_sgRNA_NaCl_vs_ars$tab.ExIn[order(res_sgRNA_NaCl_vs_ars$tab.ExIn$FDR, decreasing=FALSE),]
out_sgRNA_UV_vs_ars <- res_sgRNA_UV_vs_ars$tab.ExIn[order(res_sgRNA_UV_vs_ars$tab.ExIn$FDR, decreasing=FALSE),]

out_sgRNA_NaCl_vs_HS <- res_sgRNA_NaCl_vs_HS$tab.ExIn[order(res_sgRNA_NaCl_vs_HS$tab.ExIn$FDR, decreasing=FALSE),]
out_sgRNA_UV_vs_HS <- res_sgRNA_UV_vs_HS$tab.ExIn[order(res_sgRNA_UV_vs_HS$tab.ExIn$FDR, decreasing=FALSE),]

out_sgRNA_UV_vs_NaCl <- res_sgRNA_UV_vs_NaCl$tab.ExIn[order(res_sgRNA_UV_vs_NaCl$tab.ExIn$FDR, decreasing=FALSE),]
########

# change to write to 'sgRNA_comparisons' folder
write.table(format(out_sgRNA_HS_vs_ars, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_HS_vs_ars.tsv'), sep="\t", row.names=TRUE, quote = FALSE)
write.table(format(out_sgRNA_NaCl_vs_ars, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_NaCl_vs_ars.tsv'), sep="\t", row.names=TRUE, quote = FALSE)
write.table(format(out_sgRNA_UV_vs_ars, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_UV_vs_ars.tsv'), sep="\t", row.names=TRUE, quote = FALSE)

write.table(format(out_sgRNA_NaCl_vs_HS, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_NaCl_vs_HS.tsv'), sep="\t", row.names=TRUE, quote = FALSE)
write.table(format(out_sgRNA_UV_vs_HS, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_UV_vs_HS.tsv'), sep="\t", row.names=TRUE, quote = FALSE)

write.table(format(out_sgRNA_UV_vs_NaCl, digits=3), file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_UV_vs_NaCl.tsv'), sep="\t", row.names=TRUE, quote = FALSE)
###########

#BiocManager::install("EnsDb.Hsapiens.v86")
suppressPackageStartupMessages(library(EnsDb.Hsapiens.v86))
#supportedFilters(EnsDb.Hsapiens.v86)

# Convert from ensembl.gene to gene.symbol   
out_sgRNA_HS_vs_ars_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_sgRNA_HS_vs_ars), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
out_sgRNA_NaCl_vs_ars_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_sgRNA_NaCl_vs_ars), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
out_sgRNA_UV_vs_ars_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_sgRNA_UV_vs_ars), keytype = "GENEID", columns = c("SYMBOL","GENEID"))

out_sgRNA_NaCl_vs_HS_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_sgRNA_NaCl_vs_HS), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
out_sgRNA_UV_vs_HS_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_sgRNA_UV_vs_HS), keytype = "GENEID", columns = c("SYMBOL","GENEID"))

out_sgRNA_UV_vs_NaCl_geneSymbol <- ensembldb::select(EnsDb.Hsapiens.v86, keys=row.names(out_sgRNA_UV_vs_NaCl), keytype = "GENEID", columns = c("SYMBOL","GENEID"))
###

# change to write to 'sgRNA_comparisons' folder
write.table(out_sgRNA_HS_vs_ars_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_HS_vs_ars_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
write.table(out_sgRNA_NaCl_vs_ars_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_NaCl_vs_ars_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
write.table(out_sgRNA_UV_vs_ars_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_UV_vs_ars_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)

write.table(out_sgRNA_NaCl_vs_HS_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_NaCl_vs_HS_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
write.table(out_sgRNA_UV_vs_HS_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_UV_vs_HS_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)

write.table(out_sgRNA_UV_vs_NaCl_geneSymbol, file=paste0(path_inDir, '/intron_analysis_bam/', 'out_fC_sgRNA_UV_vs_NaCl_geneSymbol'), sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
```



## Run EISA step by step
# load the metadata (??)
```{r, check_counts, eval=FALSE}
# fmeta_all <- "sampleSheets/all_batch_corrected_sampleSheet.tsv"
# dfmeta_all <- read.delim(paste0(path_inDir, '/', fmeta_all), header = TRUE, sep='\t', na.strings="NA" , row.names=1)
# dfmeta_all$batch <- factor(dfmeta_all$batch, levels=c("collA","collB")) 
# dfmeta_all$condition <- factor(dfmeta_all$condition, levels=c("totRNA_mock","totRNA_arsenite", "totRNA_HS", "totRNA_NaCl", "totRNA_UV", "sgRNA_arsenite", "sgRNA_HS", "sgRNA_NaCl", "sgRNA_UV")) 
# dfmeta_all$group <- factor(dfmeta_all$group, levels=c("All","treatment1" ,"treatment2","treatment3" ,"treatment4")) 
```




```{r, eisa-manual, eval=FALSE}
# ## 1. Normalization
# fracIn <- colSums(counts_introns)/colSums(less_counts_geneBody)
# summary(fracIn)
# 
# # scale counts to the mean library size,
# # separately for exons and introns
# Nex <- t(t(less_counts_exon) / colSums(less_counts_exon) * mean(colSums(less_counts_exon)))
# Nin <- t(t(counts_introns) / colSums(counts_introns) * mean(colSums(counts_introns)))
# 
# # log transform (add a pseudocount of 8)
# NLex <- log2(Nex + 8)
# NLin <- log2(Nin + 8)          # WARNING: NaNs produced
# 
# #Identify quantifiable genes
# #remove very low counts in either exons or introns
# #can use a fixed cut-off on the normalized, log-transformed intron and exonic counts:
# 
# quantGenes <- rownames(less_counts_exon)[ rowMeans(NLex) > 5.0 & rowMeans(NLin) > 5.0 ]
# length(quantGenes)
# 
# #calculate the changes between neurons and ES cells in introns (ΔI), in exons (ΔE), and the difference between the two (ΔE−ΔI):
# Dex <- NLex[,c("s1","s2","s3")] - NLex[,c("s4","s5","s6")]
# Din <- NLin[,c("s1","s2","s3")] - NLin[,c("s4","s5","s6")]
# Dex.Din <- Dex - Din
# 
# cor(Dex[quantGenes,1], Dex[quantGenes,2])
# #> [1] 0.9379397
# cor(Din[quantGenes,1], Din[quantGenes,2])
# #> [1] 0.8449252
# cor(Dex.Din[quantGenes,1], Dex.Din[quantGenes,2])
# #> [1] 0.5518187
```

```{r, statistical-analysis, eval=FALSE}
# # create DGEList object with exonic and intronic counts
# library(edgeR)
# cnt <- data.frame(Ex = less_counts_exon, In = counts_introns)
# y <- DGEList(counts = cnt, genes = data.frame(ENTREZID = rownames(cnt)))
# 
# # select quantifiable genes and normalize
# y <- y[quantGenes, ]
# y <- calcNormFactors(y)
# 
# 
# 
# # design matrix with interaction term
# region <- factor(rep(c("ex","in"), each=6, levels = c("in", "ex")) )
# 
# #cond <- rep(factor(c("ES","ES","TN","TN")), 2)
# cond <- droplevels(treat[c(1:3, 4:6)])
# 
# design <- model.matrix(~ region * cond)
# rownames(design) <- colnames(cnt)
# design
# 
# # estimate model parameters
# y <- estimateDisp(y, design)
# fit <- glmFit(y, design)
# 
# # calculate likelihood-ratio between full and reduced models
# lrt <- glmLRT(fit)
# 
# # create results table
# tt <- topTags(lrt, n = nrow(y), sort.by = "none")
# head(tt$table[order(tt$table$FDR, decreasing = FALSE), ])
# ################################
```


```{r, vignette}
# #annotation
# txdbFile <- system.file("extdata", "hg19sub.sqlite", package = "eisaR")
# txdb <- AnnotationDbi::loadDb(txdbFile)
# 
# #load count data
# cntEx <- readRDS(system.file("extdata",
#                              "Fig3abc_GSE33252_rawcounts_exonic.rds",
#                              package = "eisaR"))
# cntIn <- readRDS(system.file("extdata",
#                              "Fig3abc_GSE33252_rawcounts_intronic.rds",
#                              package = "eisaR"))
# 
# #run eisa
# # remove "width" column
# Rex <- cntEx[, colnames(cntEx) != "width"]
# Rin <- cntIn[, colnames(cntIn) != "width"]
# 
# # create condition factor (contrast will be TN - ES)
# cond <- factor(c("ES", "ES", "TN", "TN"))
# 
# # run EISA
# res <- runEISA(Rex, Rin, cond)

```



```{r, eisa-ciri-comp}
# extract top 2000 genes from each comparison 

list_objects=list("totRNA_ars_vs_mock", "totRNA_HS_vs_mock", "totRNA_NaCl_vs_mock", "totRNA_UV_vs_mock",
                  "sgRNA_HS_vs_ars", "sgRNA_NaCl_vs_ars", "sgRNA_UV_vs_ars",
                   "sgRNA_NaCl_vs_HS", "sgRNA_UV_vs_HS",
                   "sgRNA_UV_vs_NaCl")
 
for (i in 1:length(list_objects) ) {
  assign(paste0("top2000_", list_objects[[i]]), get(paste0("out_", list_objects[[i]]))[1:2000,]) 
  write.table(get(paste0("top2000_", list_objects[[i]])), file=paste0(path_inDir, '/intron_analysis_bam/', "top2000_", list_objects[[i]], '.tsv'), sep="\t", row.names=TRUE, quote = FALSE)
}
```


```{r, function-eisa-plot-modif}

f.my_plotEISA <- function(x, contrast = c("ExIn", "none"),
                     minLfc = NULL, maxFDR = 0.05,
                     genecolors = c("#E41A1C", "#497AB3", "#222222"), ...) {  # meaninig = significantly up, down or unchanged
    # check arguments
    contrast <- match.arg(contrast)
    contrastName <- ifelse("contrastName" %in% names(x), paste0(" (",x$contrastName, ")"), "")
    sigtab <- switch(contrast, ExIn = x$tab.ExIn, none = data.frame())
    if (nrow(sigtab) == 0 && contrast != "none")
        stop("'x' does not contain the requested statistics and can only ",
             "be plotted using contrast = 'none'. Note that at least two ",
             "replicates per condition are required to run the statistical ",
             "testing.")
    if (is.null(minLfc))
        minLfc <- 0
    stopifnot(exprs = {
        is.numeric(minLfc)
        length(minLfc) == 1L
        is.numeric(maxFDR)
        length(maxFDR) == 1L
        maxFDR >= 0
        maxFDR <= 1.0
        is.character(genecolors)
        length(genecolors) == 3L
    })

    # identify gene to highlight
    if (contrast == "none") {
        sig <- rep(FALSE, nrow(x$contrasts))
        sig.dir <- numeric(0)
    } else {
        sig <- abs(sigtab$logFC) >= minLfc & sigtab$FDR <= maxFDR
        sig.dir <- sign(sigtab$logFC[sig])
        message("identified ", sum(sig), " genes to highlight")
    }

    # set graphical parameters (user-defined colors take precedence)
    dotsL <- list(...)
    dotsL$x <- x$contrasts[, "Din"]
    dotsL$y <- x$contrasts[, "Dex"]
    if (!"pch" %in% names(dotsL))
        dotsL$pch <- "."
    if (!"cex" %in% names(dotsL))
        dotsL$cex <- 2
    if (!"col" %in% names(dotsL))
        dotsL$col <- ifelse(sig, ifelse(sigtab$logFC > 0, genecolors[1], genecolors[2]), genecolors[3])  # sign up, down or unchanged
    #col=ifelse(object$sig ==TRUE & object$belong ==TRUE, colEmph, ifelse(object$sig==FALSE & object$belong==TRUE, colAmbig, ifelse(object$sig==TRUE & object$belong==FALSE, colSig, colNonSig)) ),
    if (!"xlab" %in% names(dotsL))
        dotsL$xlab <- substitute(expression(paste(Delta, "intron", cn)), list(cn = contrastName))
    if (!"ylab" %in% names(dotsL))
        dotsL$ylab <- substitute(expression(paste(Delta, "exon", cn)), list(cn = contrastName))

    # Delta I vs. Delta E
    do.call(plot, dotsL)
    if (contrast != "none") {
        if (length(dotsL$col) == 1L)
            dotsL$col <- rep(dotsL$col, length(dotsL$x))
        points(dotsL$x[sig], dotsL$y[sig], pch = 20, col = dotsL$col[sig])
        legend(x = "bottomright", bty = "n", pch = 20, col = genecolors[c(1,2)],
               legend = sprintf("%s (%d)", c("Up","Down"), c(sum(sig.dir == 1), sum(sig.dir == -1))))
    }

    return(invisible(NULL))
}
```



```{r, heatmap-new-plot}
# exon counts
head(less_counts_exon)

# gene body counts
head(less_counts_geneBody)

# import genes from Yilong
genes_Yilong <- read.delim('/data/manke/processing/balan/Yilong_SG_data/2022.03_DE_genes_transcript/analysis_bam_default/new_plots/genes.tab', header = F, sep='\t', na.strings="NA")

# subset the file by finding the index of the rows starting with '#'
#genes_Yilong[grep("^#", genes_Yilong$V1), ]
indices <- grep("^#", genes_Yilong$V1)

# read each section of the file as a separate df:
section_names <-genes_Yilong[grep("^#", genes_Yilong$V1), ]$V1
# reformat names
section_names <- gsub("#","",section_names) 

section_names[2] <- gsub("-","_",section_names[2])
section_names[3] <- strsplit(section_names[3], ' ')[[1]][1]
section_names[4] <- gsub(" ","_",section_names[4])

# assign each section to names and headers
for (i in seq(1, length(indices)-1) ) {
  assign(section_names[i], genes_Yilong[c((indices[i]+2):indices[i+1]-1), ]  )
  print(dim(get(section_names[i]))[1])
}

# load DB to translate the symbols
library(biomaRt)

# symbols ids read into the 4 df:
# list available DB that exist in Ensembl:
#listEnsembl()
# using Ensembl90
listEnsemblArchives()
listEnsembl(version = 90)

ensembl90 <- useEnsembl(biomart = 'genes', 
                       dataset = 'hsapiens_gene_ensembl',
                       version = 90)

# retrieve the attributes for the query:
attr <- listAttributes(ensembl90)
# what I need is symbol -> ensemblID
#'hgnc_symbol', 'QQQexternal_gene_name', 'PPPensembl_gene_id', 'hgnc_id'
		
# retrieve the filters for the query:
filters <- listFilters(ensembl90)
# what filters I want: gene stable id -> 'ensembl_gene_id'

# build a query and run it:
df.G1_S <- getBM(attributes = c('external_gene_name', 'ensembl_gene_id'),
		filters = "external_gene_name",
		values = G1_S$V2[-1],
		mart = ensembl90)

df.G2_M <- getBM(attributes = c('external_gene_name', 'ensembl_gene_id'),
		filters = "external_gene_name",
		values = G2_M$V2[-1],
		mart = ensembl90)

df.inflammatory <- getBM(attributes = c('external_gene_name', 'ensembl_gene_id'),
		filters = "external_gene_name",
		values = Inflammatory$V1[-1],
		mart = ensembl90)

df.dsRNA_response <- getBM(attributes = c('external_gene_name', 'ensembl_gene_id'),
		filters = "external_gene_name",
		values = dsRNA_response$V1[-1],
		mart = ensembl90)


# subset genes for each df
#subset(df, subset = rownames(df) %in% genelist)
counts.G1_S <- subset(less_counts_exon, subset = row.names(less_counts_exon) %in% df.G1_S$ensembl_gene_id)
counts.G2_M <- subset(less_counts_exon, subset = row.names(less_counts_exon) %in% df.G2_M$ensembl_gene_id)
counts.inflammatory <- subset(less_counts_exon, subset = row.names(less_counts_exon) %in% df.inflammatory$ensembl_gene_id)
counts.dsRNA_response <- subset(less_counts_exon, subset = row.names(less_counts_exon) %in% df.dsRNA_response$ensembl_gene_id)

# heatmaps for each of them with all samples
heatmap(as.matrix(counts.G1_S), Rowv = NA, Colv = NA)
heatmap(as.matrix(counts.G2_M), Rowv = NA, Colv = NA)
heatmap(as.matrix(counts.inflammatory), Rowv = NA, Colv = NA)
heatmap(as.matrix(counts.dsRNA_response), Rowv = NA, Colv = NA)


# pull all of them together
counts.all_samples <- rbind(counts.G1_S, counts.G2_M, counts.inflammatory, counts.dsRNA_response)

# give a different color map

png("../2218_provisory_heatmap.png",    # create PNG for the heat map        
  width = 5*300,        # 5 x 300 pixels
  height = 5*300,
  res = 300,            # 300 pixels per inch
  pointsize = 8)        # smaller font size

my_palette <- colorRampPalette(c("blue", "white", "red"))(n = 500)
heatmap(as.matrix(counts.all_samples), Rowv = NA, Colv = NA, col=my_palette, main='2218')

dev.off()               # close the PNG device
```




```{r, plot-eisa-with-ciridata}


# import the data with top 2000 genes from the eisa analysis
path_orig <- "/data/manke/processing/balan/Yilong_SG_data/2022.03_DE_genes_transcript/intron_analysis_bam"


# #DON'T CHANGE THE ORDER OF THE OBJECTS!!!
# 
# list_objects=list(#"totRNA_ars_vs_mock", "totRNA_HS_vs_mock", "totRNA_NaCl_vs_mock", "totRNA_UV_vs_mock",
#                    "sgRNA_HS_vs_ars", "sgRNA_NaCl_vs_ars", "sgRNA_UV_vs_ars",
#                    "sgRNA_NaCl_vs_HS", "sgRNA_UV_vs_HS",
#                    "sgRNA_UV_vs_NaCl")

list_toExtract <- list("DE_gene_sgRNA_UV_vs_sgRNA_arsenite", "DE_gene_sgRNA_UV_vs_sgRNA_HS", 
                     "DE_gene_sgRNA_UV_vs_sgRNA_NaCl", "DE_gene_sgRNA_HS_vs_sgRNA_arsenite")


for (i in 1:length(list_toExtract) ) {
  assign(paste0("top2000_", list_toExtract[[i]], "_padj0.05lfc_sort"), read.delim(paste0(path_orig, '/', "top2000_", list_toExtract[[i]], "_padj0.05lfc_sort", '.tsv'), header = TRUE, sep='\t', na.strings="NA", row.names = 1, stringsAsFactors = F) )
}


# png(paste0(path_inDir, '/intron_analysis_bam/sgRNA_comparisons/','eisa_plot_sgRNA_LFC_revUV.png'), width=1600, height=800, units = "px", bg = "white")

par(oma = c(3,4,4,1),    # margins for whole plot
    mar = c(3,4,3.2,1)  # margins for individual plots
    )

xl <- c(-12,12)
yl <- c(-8,8)

layout(matrix(c(1, 2, 3, 4), 2, 3, byrow=TRUE))
# layout.show(6)

list_objects=list(#"res_sgRNA_HS_vs_ars", "res_sgRNA_NaCl_vs_HS", 
  "res_sgRNA_UV_vs_ars", "res_sgRNA_UV_vs_HS", "res_sgRNA_UV_vs_NaCl", "res_sgRNA_NaCl_vs_ars")


for (i in 1:length(list_objects) ) {
  f.my_plotEISA(get(list_objects[[i]]), maxFDR = 0.05, minLfc = 2,
                  xlim=xl, ylim=yl,
                  main=list_objects[[i]] , xlab='', ylab='',
                  cex.axis=1.5, cex.main=1.5)
  abline(h = 0, v = 0, col = "gray")
}

title(main = "EISA - sgRNA_intron-exon analysis, FDR < 0.05, LFC > 2",
      xlab = expression(paste(Delta,'intron (cond1-cond2)')),
      ylab = expression(paste(Delta,'exon (cond1-cond2)')),
      outer = TRUE, line = 1,
      cex.lab=2, cex.main=2)

# 
# dev.off()
```





###########################

